docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,398 @@
+- Implementation details: Only if you want to understand the current implementation at a high level.
+
+
+
+Mathematical formulation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+What does the autograd package do
+"""""""""""""""""""""""""""""""""
+
+The autograd package allows to use automatic differentiation techniques on top of PyTorch Tensors.
+Automatic differentiation allows to compute dot product between a given vector and the Jacobian of a user-defined function.
+In this document, we will consider the function :math:`f: x \rightarrow y` where :math:`x` is a vector of size :math:`I` and :math:`y` is a vector of size :math:`O`.
+For simplicity, in this document, we will always consider inputs and ouputs to be 1D, in practice, they can be Tensors with an arbitrary number of dimensions.
+The jacobian matrix associated with this function is that matrix :math:`J_f` of size :math:`O \times I` such that each entry is
+given by :math:`(J_f)_{ij} = \dfrac{\partial{x_j}}{\partial{y_i}}`. For simplicity, we will write :math:`J_f = \dfrac{\partial{y}}{\partial{x}}`.
+
+Currently, PyTorch only implements reverse mode automatic differentiation which, given
+an arbitrary vector :math:`v` of size :math:`O`, will compute :math:`v^T J_f`.
+
+Note the particularly interesting case where :math:`f`'s output is a scalar value because providing a vector :math:`v` containing
+the value :math:`1` will compute exactly :math:`J_f`.
+This is used extensively in deep learning for example where loss functions are scalar and reverse mode automatic
+differentiation is used to compute the gradients of the loss with respect to the weights of the model.
+It is also regularly called the backpropagation algorithm in this case.
 
-Every Tensor has a flag: :attr:`requires_grad` that allows for fine grained
-exclusion of subgraphs from gradient computation and can increase efficiency.
 
-.. _excluding-requires_grad:
 
-``requires_grad``
-~~~~~~~~~~~~~~~~~
+How does it does that
+"""""""""""""""""""""
 
-If there's a single input to an operation that requires gradient, its output
-will also require gradient. Conversely, only if all inputs don't require
-gradient, the output also won't require it. Backward computation is never
-performed in the subgraphs, where all Tensors didn't require gradients.
+One simple interpretation of what it is doing is splitting the user function :math:`f` into composition of smaller
+pre-defined operations.
+We will call such operations elementary.
+The gradient of the original function is the computed by using the chain rule on each of these elementary operations.
+
+For example, if we split :math:`f` into two elementary operations, namely :math:`y = op_2(op_1(x))`,
+the automatic differentiation will compute :math:`v^T J_f = (v^T J_{op_2}) J_{op_1} =  v^T J_{op_2} J_{op_1}`.
+
+Link with the code
+""""""""""""""""""
+
+The tensors for which the user want to compute the gradients can be marked using the :code:`.requires_grad_()` function.
```

Comment:
You've lost some content from the original text. One of the clever things about `requires_grad` is that it allows us to bypass tracking of the autograd history when none of the inputs were declared to require gradients. This is what the original text spends an entire section talking about, and it's totally glossed over here.

Reply:
Also, using `requires_grad_` to inplace mark tensors as requiring gradients is not that idiomatic, right? Usually you'd rather just say something like `torch.zeros(x, requires_grad=True)`